fw/str.c

```diff
@@ -670,6 +688,139 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
+ *          nchunks = 2
+ */
+int
+tfw_str_array_append_chunk(TfwPool *pool, TfwStr *array,
+			   char *data, unsigned long len,
+			   bool complete_last)
+{
+	TfwStr *last_token = NULL, *last_chunk;
+	unsigned long added_len = 0;
+	bool shall_complete;
+
+	/* Quick return when there's nothing to do */
+	if (!len && !complete_last)
+		return 0;
+
+	BUG_ON(!data);
+	if (array->nchunks)
+		last_token = &array->chunks[array->nchunks - 1];
+	/*
+	 * Only try to complete an existing uncompleted item or to
+	 * the one newly created at the below.
+	 */
+	if ((len || (last_token && !(last_token->flags & TFW_STR_COMPLETE)))
+	    && complete_last)
+	{
+		shall_complete = true;
+	} else {
+		shall_complete = false;
+	}
+
+	if (!len) {
+		/* No chunks to be appended, just mark the item as completed */
+		if (shall_complete) /* implies last_token != NULL */
+			last_token->flags |= TFW_STR_COMPLETE;
+
+		return 0;
+	}
+
+	/*
+	 * Allocate memory required for the new token and/or chunks.
+	 * tfw_pool_realloc() is called at most once: either to reallocate the
+	 * array items or reallocate an item chunks. In the first case the pool
+	 * can be called once more to allocate chunks for a new item, but there
+	 * will be only one trivial TfwPool->off movement.
+	 *
+	 * Empty chunks array is always initialized with zeroes.
+	 *
+	 * Step 1: allocate a new uncompleted first-level item if none exists or
+	 * if the last array item is completed.
+	 */
+	if (!last_token) { /* implies array->nchunks == 0 */
+		/*
+		 * Cannot store "last_token" at the place of "array" the way
+		 * "__str_grow_tree" implies, because we need distinct nchunks
+		 * for different levels of the tree.
+		 */
+		last_token = __str_alloc(pool, array, 1);
+		if (unlikely(!last_token))
+			return -ENOMEM;
+	}
+	else if (last_token->flags & TFW_STR_COMPLETE) {
+		/*
+		 * If the last token is complete, then we need to add some more
+		 * data next.
+		 */
+		if (unlikely(array->nchunks == __TFW_STR_ARRAY_MAX))
+			return -E2BIG;
+		last_token = __str_grow_tree(pool, array, 0, 1);
+		if (unlikely(!last_token)) {
+			/*
+			 * Just return an error code and do not call
+			 * tfw_pool_free(). The caller won't proceed with the
+			 * failure and will block the HTTP message, so the whole
+			 * pool is going to be freed soon.
+			 */
+			return -ENOMEM;
+		}
+	}
+	/*
+	 * Step 2: append second-level chunks into the uncompleted token/item.
+	 * last_token is empty only if it was just allocated on Step 1, so only
+	 * a trivial TfwPool->off movement happens in this case.
+	 */
+	if (TFW_STR_EMPTY(last_token)) {
+		/* Merge single item into parent to avoid reallocation */
+		last_chunk = last_token;
+	} else {
+		last_chunk = __str_grow_tree(pool, last_token, 0, 1);
+	}
+
+	if (!last_chunk)
+		return -ENOMEM;
+
+	added_len += len;
+	last_chunk->len = len;
+	last_chunk->data = data;
+	/*
+	 * We don't care about array->len, but the last_token->len is required
```

Comment:
I think it makes sense to replace
```
We don't care about array->len, but
```
with
```
array->len has no sense since this is an array of independent strings. However,
```